freqtrade/tests/edge/test_edge.py

```diff
@@ -0,0 +1,275 @@
+from freqtrade.tests.conftest import get_patched_exchange
+from freqtrade.edge import Edge
+from pandas import DataFrame, to_datetime
+from freqtrade.strategy.interface import SellType
+import arrow
+import numpy as np
+import math
+
+from unittest.mock import MagicMock
+
+
+# Cases to be tested:
+# 1) Open trade should be removed from the end
+# 2) Two complete trades within dataframe (with sell hit for all)
+# 3) Entered, sl 1%, candle drops 8% => Trade closed, 1% loss
+# 4) Entered, sl 3%, candle drops 4%, recovers to 1% => Trade closed, 3% loss
+# 5) Entered, sl 2%, candle drops 4%, recovers to 1%, entry met, candle drops 20% =>
+#    Trade 1 closed: loss 2%, Trade 2 opened, Trade 2 closed: loss 2%
+# 6)
+###################################################################
+# STOPLOSS:
+# 6) Candle drops 8%, stoploss at 1%: Trade closed, 1% loss
+# 7) Candle drops 4% but recovers to 1% loss, stoploss at 3%: Trade closed, 3% loss
+# 8) Candle drops 4% recovers to 1% entry criteria are met, candle drops
+#    20%, stoploss at 2%: Trade 1 closed, Loss 2%, Trade 2 opened, Trade 2 closed, Loss 2%
+####################################################################
+# PRIORITY TO STOPLOSS:
+# 9) Stoploss and sell are hit. should sell on stoploss
+####################################################################
+
+ticker_start_time = arrow.get(2018, 10, 3)
+ticker_interval_in_minute = 60
+_ohlc = {'date': 0, 'buy': 1, 'open': 2, 'high': 3, 'low': 4, 'close': 5, 'sell': 6, 'volume': 7}
+
+
+def test_filter(mocker, default_conf):
+    exchange = get_patched_exchange(mocker, default_conf)
+    edge = Edge(default_conf, exchange)
+    mocker.patch('freqtrade.edge.Edge._cached_pairs', mocker.PropertyMock(
+        return_value=[
+            ['E/F', -0.01, 0.66, 3.71, 0.50, 1.71],
+            ['C/D', -0.01, 0.66, 3.71, 0.50, 1.71],
+            ['N/O', -0.01, 0.66, 3.71, 0.50, 1.71]
+        ]
+    ))
+
+    pairs = ['A/B', 'C/D', 'E/F', 'G/H']
+    assert(edge.filter(pairs) == ['E/F', 'C/D'])
+
+
+def test_stoploss(mocker, default_conf):
+    exchange = get_patched_exchange(mocker, default_conf)
+    edge = Edge(default_conf, exchange)
+    mocker.patch('freqtrade.edge.Edge._cached_pairs', mocker.PropertyMock(
+        return_value=[
+            ['E/F', -0.01, 0.66, 3.71, 0.50, 1.71],
+            ['C/D', -0.01, 0.66, 3.71, 0.50, 1.71],
+            ['N/O', -0.01, 0.66, 3.71, 0.50, 1.71]
+        ]
+    ))
+
+    assert edge.stoploss('E/F') == -0.01
+
+
+def _validate_ohlc(buy_ohlc_sell_matrice):
+    for index, ohlc in enumerate(buy_ohlc_sell_matrice):
+        # if not high < open < low or not high < close < low
+        if not ohlc[3] > ohlc[2] > ohlc[4] or not ohlc[3] > ohlc[5] > ohlc[4]:
```

Comment:
i'm not 100% certain, but i think you should check `>=` here - `high` can be identical to open / close ... 
it doesn't necessarily pull back at the end of the candle ...  (same with low / close / ... )

Reply:
Eagle eye. will be corrected.